pkg/oci.go

```diff
@@ -74,48 +71,66 @@ func PushSBOM(sbomDoc *v2_3.Document, sbomDescriptor *v1.Descriptor, sbomBytes [
 	})
 	if err != nil {
-		return fmt.Errorf("error packing manifest: %w", err)
+		return nil, fmt.Errorf("error packing manifest: %w", err)
 	}
 
 	// Use the latest tag if no tag is specified
 	tag := "latest"
 	ref, err := registry.ParseReference(reference)
 	if err != nil {
-		return fmt.Errorf("error parsing reference: %w", err)
+		return nil, fmt.Errorf("error parsing reference: %w", err)
 	}
 
 	if ref.Reference != "" {
 		tag = ref.Reference
 	}
 
 	if err = mem.Tag(ctx, manifestDescriptor, tag); err != nil {
-		return err
+		return nil, err
 	}
 
-	// Connect to a remote repository
-	repo, err := remote.NewRepository(reference)
+	// Copy from the memory store to the remote repository
+	manifest, err := oras.Copy(ctx, mem, tag, dest, tag, oras.DefaultCopyOptions)
+	return &manifest, err
+}
+
+// AttachArtifact attaches an artifact to the subject descriptor
+func AttachArtifact(subject *v1.Descriptor, artifactDescriptor *v1.Descriptor, artifactType string, artifactBytes []byte, dest oras.Target) (*v1.Descriptor, error) {
+	ctx := context.Background()
+
+	exists, err := dest.Exists(ctx, *subject)
 	if err != nil {
-		return fmt.Errorf("error connecting to remote repository: %w", err)
+		return nil, fmt.Errorf("error checking if subject exists in dest: %w", err)
 	}
-
-	// Check if registry has is localhost or starts with localhost:
-	reg := repo.Reference.Registry
-	if strings.HasPrefix(reg, "localhost:") {
-		repo.PlainHTTP = true
+	if !exists {
+		return nil, fmt.Errorf("subject descriptor does not exist in dest")
 	}
 
-	// Prepare the auth client for the registry
-	client := &auth.Client{
-		Client: retry.DefaultClient,
-		Cache:  auth.DefaultCache,
+	// Create a Reader for the bytes
+	artifactReader := bytes.NewReader(artifactBytes)
+
+	// Add descriptor to a memory store
+	mem := memory.New()
+	err = mem.Push(ctx, *artifactDescriptor, artifactReader)
+	if err != nil {
+		return nil, fmt.Errorf("error pushing artifact into memory store: %w", err)
 	}
 
-	client.Credential = credsResolver
+	// Pack the artifact manifest with the subject descriptor
+	artifactManifestDescriptor, err := oras.PackManifest(ctx, mem, oras.PackManifestVersion1_1, artifactType, oras.PackManifestOptions{
+		Subject: subject,
+		Layers:  []v1.Descriptor{*artifactDescriptor},
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("error packing artifact manifest: %w", err)
+	}
 
-	client.SetUserAgent(APPLICATION_USERAGENT)
-	repo.Client = client
+	// Tag the artifact manifest with the artifact digest
+	if err = mem.Tag(ctx, artifactManifestDescriptor, artifactManifestDescriptor.Digest.String()); err != nil {
+		return nil, fmt.Errorf("error tagging artifact: %w", err)
+	}
 
-	// Copy from the memory store to the remote repository
-	_, err = oras.Copy(ctx, mem, tag, repo, tag, oras.DefaultCopyOptions)
-	return err
+	manifest, err := oras.Copy(ctx, mem, artifactManifestDescriptor.Digest.String(), dest, artifactManifestDescriptor.Digest.String(), oras.DefaultCopyOptions)
```

Comment:
We don't need to tag the descriptor by digest. We can just use [`oras.CopyGraph`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0#CopyGraph) which does not require a tag.